### ISSUE TITLE
Document behaviour of unnamed expressions in the SELECT clause

### ISSUE THREAD

**Issue by McNetic** [jOOQ/jOOQ#7968]
### Expected behavior and actual behavior:
First of all, I'm not sure this should really be considered a bug or rather a combination of unfortunately chosen settings and poor SQL skills (or lack of knowledge of best practices) on our side. But on topic:

We chose using RenderNameStyle.LOWER (with the default QUOTED the problem does not arise).

Usually, I expect jooq queries (using standard features) to produce valid sql regardless of chosen SQL dialect (provided that the features are supported by that dialect).

It is however possible to write queries that work with oracle 12 dialect, but don't with oracle 11, because of keywords being choosen by jOOQ as aliases for subqueries (and perhaps other expressions, too).

The problem can be fixed (or worked around) by manually chosing aliases for said expressions. 

### Steps to reproduce the problem (include example code if possible):
One simple example of a problematic query is the following:
```
select(
  select(inline("a")).from(dual()).asField()
)
.from(dual())
.limit(0, 10)
```
In Oracle 12, this is rendered with `fetch next 10 rows`. In Oracle 11, the `limit()` method has to be emulated. The generated query is then:
```
SELECT v0 select
FROM (
  SELECT 
    x.v0, 
    rownum rn
  FROM (
    SELECT (
      SELECT 'b'
      FROM (
        SELECT 'X' dummy
        FROM dual
      ) dual
    ) v0
    FROM (
      SELECT 'X' dummy
      FROM dual
    ) dual
  ) x
  WHERE rownum <= (0 + 10)
)
WHERE rn > 0
```
This breaks on the alias `select` in the first line. It works again as soon as I use `asField('name')` for the subselect.

I would at least expect the manual (https://www.jooq.org/doc/3.0/manual/sql-building/dsl-context/custom-settings/settings-name-style/) telling explicitly that using anything but QUOTED may not be a good idea. Additionally, you might consider preventing jOOQ from autogenerating keyword aliases.

For the best practices:
- should I ever use anything but RenderNameStyle.QUOTED?
- should I always manually set aliases for subqueries / query parts?

### Versions:

- jOOQ: 3.11.5
- Java: 8
- Database (include vendor): Oracle 11/12
- OS: Windows
- JDBC Driver (include name if inofficial driver): ojdbc7

**Comment by lukaseder** [comment]
Thank you very much for your report. That's unfortunate - your API usage doesn't seem wrong to me, even if providing an alias would probably be recommended, it shouldn't be necessary.

> I would at least expect the manual (https://www.jooq.org/doc/3.0/manual/sql-building/dsl-context/custom-settings/settings-name-style/) telling explicitly that using anything but QUOTED may not be a good idea. 

This is just a bug, it doesn't mean that using the other `RenderNameStyle` flags is a good idea.

> should I ever use anything but RenderNameStyle.QUOTED?

The default is QUOTED because that will prevent most problems, but people might prefer not to have these quotes.

> should I always manually set aliases for subqueries / query parts?

That depends on how you're using jOOQ's generated SQL. jOOQ tends to generate aliases when needed to make the SQL work, so they're not strictly required, although when you write a correlated subquery in SQL (or any "complex" expression in the `SELECT` clause), it is definitely not unusual to provide an alias.

**Comment by lukaseder** [comment]
After more careful review, I would say this is not strictly a bug, but indeed an undocumented "feature". Currently, unnamed queryparts will produce arbitrary names that can cause a variety of issues, including:

- #579 SQL syntax error when using DSL.selectOne() in a subquery

Another, related problem is that sometimes perfectly valid ambiguous column names of named queryparts in a top level query will no longer be valid if jOOQ nests those columns in derived tables. Example issues:

- #7867 DSLContext.fetchCount(Select) should rename the select statement's column names to prevent ambiguities 
- #7877 Ambiguous record warning thrown while using implicit join syntax on tables which have a common column name

### Thorough solution to all of this

A thorough solution to all of this might be to define a sound strategy that disambiguates all of these column names throughout a query by renaming them to some synthetic name in a subquery, and, possibly, back to the original in the top level query. I've created a new feature request for this: #7977

One of the difficulties here is that any ordinary `Field` extends `SelectField` extends `Named` for convenience reasons, so every field has to statically provide some sort of name (e.g. `select`), which may only be usable inside of quotes.

### Quick fix for your immediate problem

So, another solution here is to define an additional `RenderNameStyle` that quotes names only if "needed". This is what other ORMs do, as they tend not to default to quoting everything, with the risk of being incomplete on some new, untested RDBMS, and taking some more time to generate the SQL because of the additional checks. I've created another feature request for this: #5909

I think #5909 will be a much easier and more thorough fix for you. You can have lower case identifiers *except* when they're keywords, in case of which they will use their original case and are quoted.

### This issue

I have renamed this issue to be about the documentation problem. We should better document:

- The fact that all `Field` expressions have a name (even if it is not strictly defined), and what that name does
- What the different render name styles mean
- The benefits of aliasing, and the fact that it isn't always required

Does that sound useful / have I missed something?

**Comment by lukaseder** [comment]
This task has been implemented for jOOQ 3.20:

- https://github.com/jOOQ/jOOQ/issues/579

I'll look into this one as well:

- https://github.com/jOOQ/jOOQ/issues/7977
